scripts/tools/cancel_workflows_for_pr.py

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2022 Project CHIP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+
+import click
+import coloredlogs
+from github import Github
+
+__LOG_LEVELS__ = {
+    "debug": logging.DEBUG,
+    "info": logging.INFO,
+    "warn": logging.WARN,
+    "fatal": logging.FATAL,
+}
+
+REPOSITORY = "project-chip/connectedhomeip"
+
+
+class Canceller:
+    def __init__(self, token):
+        self.api = Github(token)
+        self.repo = self.api.get_repo(REPOSITORY)
+
+    def cancel_all_runs(self, pr_number, commit_sha, dry_run):
+        pr = self.repo.get_pull(pr_number)
+        logging.info("Examining PR '%s'", pr.title)
+        for commit in pr.get_commits():
+            if commit.sha != commit_sha:
+                logging.info("Skipping SHA '%s' as it was not selected", commit.sha)
+                continue
+
+            for check_suite in commit.get_check_suites():
+                for run in check_suite.get_check_runs():
+                    if run.status in {"in_progress", "queued"}:
+                        if dry_run:
+                            logging.warning("DRY RUN: Will not stop run %s", run.name)
+                        else:
+                            logging.warning("Stopping run %s", run.name)
+                            self.repo.get_workflow_run(run.id).cancel()
+                    else:
+                        logging.info("Skip over run %s (%s)", run.name, run.status)
+
+
+@click.command()
+@click.option(
+    "--log-level",
+    default="INFO",
+    type=click.Choice(list(__LOG_LEVELS__.keys()), case_sensitive=False),
+    help="Determines the verbosity of script output.",
+)
+@click.option("--pull-request", type=int, help="Pull request number to consider")
+@click.option("--commit-sha", help="Commit to look at when cancelling pull requests")
```

Comment:
I think that this should be the HEAD commit. Otherwise, we might have a race where someone will push new changes, but this script will kill workflows because of not-killed-yet-by-github run failure, because given SHA is in the list of commits (it's not the last though).

Reply:
Github seems to associate checks with individual commits. My thinking was that we want to cancel workflow runs for the commit that failed but not cancel others. Situation like this:

- bad commit starts CI  (SHA A)
- developer notices bad commit and submits a fix (SHA B)
- new commit starts CI
- CI in SHA A fails and triggers "cancel"

in the above I want SHA-A to only cancel workflow runs for A and not for B.
Does this seem reasonable? what scenarios did you have in mind?

Reply:
> Github seems to associate checks with individual commits.

OK I see, that might have sense.